HACKING-URL.md

```diff
@@ -0,0 +1,319 @@
+add a new one, in the same format as the existing ones: the key should be
+a sensible descriptor for the relationship type, and the value an object
+containing `entity: mbid` pairs.
+
+### Entity-wide rules
+
+In some cases, you might not want to specifically match an URL to a
+relationship type, but just reject it completely for a particular entity type.
+For example, we have decided to completely disallow Wikipedia links to
+releases, since those generally belong on the release group but users often
+added them to releases using whatever relationship seemed the least bad fit.
+We now block those links on releases *and* provide an error message that
+explicitly asks the user to add them to the release group instead.
+
+To do this, you will want to use `entitySpecificRules`. For each entity type,
+`entitySpecificRules` will map to a function which can be used to reject
+links. The function works similarly to `validate` in `CLEANUPS`, and returns
+the same kind of error objects. If the entity type you want is not being used
+in `entitySpecificRules` yet, just add a new `entitySpecificRules.entity_type`
+function based on the existing ones.
+
+URL display handlers
+--------------------
+
+### Favicons
+
+For a favicon to be displayed by the URLs of the domain you are adding, you
+should add the file as a PNG (preferably sized 32x32, but 16x16 is also
+acceptable if that’s all you can find) to
+[`root/static/images/external-favicons`](root/static/images/external-favicons).
+Try to just extract the favicon directly from the site source code and convert
+it as needed, unless the site officially provides the files for download.
+
+For URLs to actually be mapped to the favicon you’ve added, you also
+need to add the mapping to `FAVICON_CLASSES` in
+[`root/static/scripts/common/constants.js`](root/static/scripts/common/constants.js).
+Then add the favicon to the list in
+[`root/static/styles/favicons.less`](root/static/styles/favicons.less):
+use the favicon class you added in the previous step, and pass a second
+argument `32` if the favicon file is 32x32 rather than 16x16.
+
+### Sidebar display
+
+:warning:
+The MVC pattern is bent here as the names displayed for hyperlinks are coded in
+the entity model layer. Plans are to address this issue tracked with the ticket
+[MBS-10605](https://tickets.metabrainz.org/browse/MBS-10605).
+
+For a domain to be displayed on the sidebar, you’ll need to create an URL model
+in [`lib/MusicBrainz/Server/Entity/URL/`](lib/MusicBrainz/Server/Entity/URL/).
+Base your model on any of the existing ones that are known to go by the same
+types of relationship, and name it after the site in question.
+
+The `sidebar_name` method most often returns a simple untranslated string,
+the name of the site. You can also return a translated string, most often when:
+* The name of the site has official localizations, for example `l(Niconico)`;
+* The site is mainly about streaming, for example `l('Stream at YouTube Music')`;
```

Comment:
I'd argue this is more about when we specifically want to specify what you'd go to the page for, but it's strange since we don't say "Buy at $store" (maybe we should!). The whole thing is weird and should either never be done or always be done IMO, but maybe you're right (re: the warning below) that just grouping these with "stream at" instead would be good enough.

Reply:
It seems that you answered your own interrogation already.